crates/astria-composer/src/searcher/collector.rs

```diff
@@ -0,0 +1,221 @@
+        url: String,
+        searcher_channel: Sender<Transaction>,
+    ) -> eyre::Result<Self> {
+        let client = EthClient::connect(&url)
+            .await
+            .wrap_err("failed connecting to eth")?;
+        let (status, _) = watch::channel(Status::new());
+        Ok(Self {
+            chain_id,
+            client,
+            searcher_channel,
+            status,
+        })
+    }
+
+    /// Subscribe to the collector's status.
+    pub(super) fn subscribe(&self) -> watch::Receiver<Status> {
+        self.status.subscribe()
+    }
+
+    #[instrument(skip_all, fields(chain_id = self.chain_id))]
+    pub(super) async fn run_until_stopped(self) -> eyre::Result<()> {
+        use std::time::Duration;
+
+        use ethers::providers::Middleware as _;
+        use futures::stream::StreamExt as _;
+        self.wait_for_evm(5, Duration::from_secs(5), 2.0)
+            .await
+            .wrap_err("failed connecting ")?;
+
+        let Self {
+            chain_id,
+            client,
+            searcher_channel,
+            ..
+        } = self;
+
+        let mut tx_stream = client
+            .inner
+            .subscribe_full_pending_txs()
+            .await
+            .wrap_err("failed to subscribe eth client to full pending transactions")?;
+        while let Some(tx) = tx_stream.next().await {
+            match searcher_channel
+                .send_timeout(
+                    Transaction {
+                        chain_id: chain_id.clone(),
+                        inner: tx,
+                    },
+                    Duration::from_millis(500),
+                )
+                .await
+            {
+                Ok(()) => {}
+                Err(SendTimeoutError::Timeout(_tx)) => {
+                    // TODO(superfluffy): what parts of tx should we log? just its hash?
+                    warn!("timed out sending new transaction to searcher after 500ms; dropping tx");
+                }
+                Err(SendTimeoutError::Closed(_tx)) => {
+                    // TODO(superfluffy): what parts of tx should we log? just its hash?
+                    warn!(
+                        "searcher channel closed while sending transaction; dropping transaction \
+                         and exiting event loop"
+                    );
+                    break;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Wait until a connection to eth is established.
+    ///
+    /// This function tries to establish a connection to eth by
+    /// querying its net_version RPC. If it fails, it retries for another `n_retries`
+    /// times with exponential backoff.
+    ///
+    /// # Errors
+    ///
+    /// An error is returned if calling eth failed after `n_retries + 1` times.
+    #[instrument(skip_all, fields(
+    retries.max_number = n_retries,
+    retries.initial_delay = %format_duration(delay),
+    retries.exponential_factor = factor,
+))]
+    async fn wait_for_evm(
```

Comment:
can we rename this to `wait_for_eth_client`?

Reply:
This function isn't waiting for the client. It's waiting for the server to respond with a certain message.